node/forest_libp2p/bitswap/tests/go_compat.rs

```diff
@@ -0,0 +1,163 @@
+                                                        .send_dont_have(true),
+                                                );
+                                            }
+                                        }
+                                        BitswapMessage::Response(cid, ..) => {
+                                            // 2. Check inbound response
+                                            if cid == outbound_request_cid {
+                                                inbound_response_tx.send_async(()).await.unwrap();
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                            RequestResponseMessage::Response { .. } => {}
+                        }
+                    }
+                    SwarmEvent::NewListenAddr {
+                        listener_id: _,
+                        address,
+                    } => {
+                        local_addr_tx.send_async(address).await.unwrap();
+                    }
+                    _ => {}
+                }
+            }
+        });
+
+        let (cancellation_tx, cancellation_rx) = flume::bounded(1);
+
+        let listen_addr = tokio::time::timeout(TIMEOUT, local_addr_rx.recv_async()).await??;
+        // Prepare `go-bitswap` test app
+        prepare_go_bitswap()?;
+        // Build and run `go-bitswap` test app
+        std::thread::spawn(move || {
+            let addr = format!("{listen_addr}/p2p/{peer_id}");
+            run_go_bitswap(cancellation_rx, addr, expected_inbound_request_cid_str)
+        });
+        let cancellation_tx_cloned = cancellation_tx.clone();
+        tokio::spawn(async move {
+            tokio::time::sleep(TIMEOUT).await;
+            println!("cancelling");
+            cancellation_tx_cloned.send_async(()).await.unwrap();
+        });
+
+        // 1. Receive request from `go-bitswap`
+        tokio::time::timeout(TIMEOUT, inbound_request_rx.recv_async()).await??;
+        println!("Received request from go-bitswap test app");
+        // 2. Receive response from `go-bitswap`
+        tokio::time::timeout(TIMEOUT, inbound_response_rx.recv_async()).await??;
+        println!("Received response from go-bitswap test app");
+        cancellation_tx.send_async(()).await.unwrap();
+
+        Ok(())
+    }
+
+    fn prepare_go_bitswap() -> Result<()> {
+        Command::new("go")
+            .args(["mod", "vendor"])
+            .current_dir(GO_APP_DIR)
+            .spawn()?
+            .wait()?;
+        Ok(())
+    }
+
+    fn run_go_bitswap(
+        cancellation_rx: flume::Receiver<()>,
+        addr: impl AsRef<str>,
+        cid: impl AsRef<str>,
+    ) -> Result<bool> {
+        let mut app = Command::new("go")
+            .args(["run", ".", "--addr", addr.as_ref(), "--cid", cid.as_ref()])
```

Comment:
It'd be helpful to put this as a test dependency (and maybe pack it into a docker image?)

Reply:
>  put this as a test dependency
Could u elaborate? 

> maybe pack it into a docker image
That will require `docker` instead of `go` to be installed, and it requires additional network setup to make 2 libp2p nodes talk to each other, maybe over complicated?

Reply:
As of this PR, we need `go` as a dependency for Forest development. This needs to be stated explicitly. A potential issue is that the `go` version is not specified, which may, for no apparent reason, fail the test in the future - that's why I'd prefer it dockerized with explicit versioning (and not `:latest`).

Putting those things aside, it may be worth putting this test under a feature flag in case I want to run the tests without installing `go` on my, e.g., new VPS.

The pains I mentioned can be solved by putting this `bitswap` crate into a separate repository with its proper CI. It wouldn't pollute Forest development dependencies. What do you think @hanabi1224?

Reply:
> go version is not specified, which may, for no apparent reason, fail the test in the future

go version >= 1.17 is specified in `go.mod` with all pinned dependency versions in `go.sum`, and it's not likely to get compilation errors unless go 2 comes out, in my experience go compiler maintains very good backward compatibility by having very limited language features

> putting this bitswap crate into a separate repository with its proper CI.

We could track this with the crate publishing story. It seems to be a good idea to put it in a separate repo once we decide to publish it, for now it makes things much easier to keep it here to void dependency version mismatch issues (namely libp2p, cid, libipld).

Reply:
Let's make sure this lands in a separate repo as soon as possible. Hidden dependencies such as this are not good and they are bad for the developer experience. It'd be useful to add `go` to the Forest dependencies in the README.

Reply:
Sure, since the absence of `go` leads to only test errors, I will add a verbose error message to this test including a link to the `go` installation instruction. It might be somehow misleading to add `go` to readme for users who only want to run `make install` . Also, the crate will likely be moved out of this repo at some point, does that sound reasonable?

Reply:
Message added

Reply:
@hanabi1224 Yeah, sure. Let's ensure it lands in a separate repo sooner than later :). In the new repo, I'd still pack the go app in a container and push it to the container registry to avoid any compilations during tests, but it's up to you; it's just a preference.